libcudacxx/include/cuda/std/__cuda/semaphore.h

```diff
@@ -38,8 +52,152 @@ class counting_semaphore : public _CUDA_VSTD::__semaphore_base<__least_max_value
+        if (limit >= cur_tickets + 1)
+        {
+            auto prev = ticketing.fetch_add(1);
+            while (!our_turn(prev + 1))
+            {
+                simple_backoff();
+            }
+        }
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY bool __acquire_slow_timed(
+        _CUDA_VSTD::chrono::nanoseconds const &__rel_time)
+    {
+        return __libcpp_thread_poll_with_backoff(
+            [this]()
+            {
+                ptrdiff_t ctr = counter.load(cuda::memory_order_relaxed);
+                while (!try_acq_impl(ctr))
+                {
+                    counter.wait(ctr);
+                }
+                int32_t cur_tickets = (ticketing.load(cuda::memory_order_relaxed) & 0xFFFFFFFF);
+                auto limit = _CUDA_VSTD::numeric_limits<int32_t>::max();
+                if (limit >= cur_tickets + 1)
+                {
+                    auto prev = ticketing.fetch_add(1);
+                    while (!our_turn(prev + 1))
+                    {
+                        simple_backoff();
+                    }
+                    return true;
+                }
+                else
+                {
+                    return false;
+                }
+            },
+            __rel_time);
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY void release(ptrdiff_t update = 1)
+    {
+        counter.fetch_add(1);
+        ticketing.fetch_add(static_cast<ptrdiff_t>(1) << 32, cuda::memory_order_relaxed);
+        // we need to notify all because no guarantee of who
+        // gets it
+        counter.notify_all();
+        ticketing.notify_all();
+    }
+
+    template <class Rep, class Period>
+    _LIBCUDACXX_INLINE_VISIBILITY bool try_acquire_for(_CUDA_VSTD::chrono::duration<Rep, Period> const &__rel_time)
+    {
+        bool result = __acquire_slow_timed(__rel_time);
+        if (result)
+        {
+            return true;
+        }
+        else
+        {
+            skipped_tickets.fetch_add(1, cuda::memory_order_relaxed);
+            return false;
+        }
+    }
+
+    template <class Clock, class Duration>
+    _LIBCUDACXX_INLINE_VISIBILITY bool try_acquire_until(_CUDA_VSTD::chrono::time_point<Clock, Duration> const &__abs_time)
+    {
+        bool result = __acquire_slow_timed(__abs_time - Clock::now());
+        if (result)
+        {
+            return true;
+        }
+        else
+        {
+            skipped_tickets.fetch_add(1, cuda::memory_order_relaxed);
+            return false;
+        }
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY constexpr explicit binary_semaphore(ptrdiff_t ctr = 1) : counter(ctr), ticketing(0)
+    {
+        assert(ctr == 1);
+    }
+
+    ~binary_semaphore() = default;
+    binary_semaphore(const binary_semaphore &) = delete;
+    binary_semaphore operator=(const binary_semaphore &) = delete;
+
+private:
+    cuda::atomic<ptrdiff_t, _Sco> counter;
+    cuda::atomic<ptrdiff_t, _Sco> skipped_tickets;
+    cuda::atomic<ptrdiff_t, _Sco> ticketing;
```

Comment:
Critical: We do not need 3 variable for a binary semaphore. As discussed offline, please start with a simple binary case first.

Reply:
The third atomic variable is there for `try_acquire_for` and `try_acquire_until`, if you want threads to have FIFO access, we need to mitigate the fact that these two methods have added to the queue. Otherwise it will hang forever if any thread uses this